### PR TITLE
AK+Kernel+Userland+Documentation: Add AK::AssertSize, start documenting patterns

### PR DESCRIPTION
**Pull request by bgianfo**
- **Documentation: Add Patterns.md**

    The purpose of this document is to track and describe the various
    patterns used through the SerenityOS code base.

- **AK: Add AssertSize utility template to provide rich type size assertions**

    This type is useful, as the sizes will be visible in the compiler error
    messages, as they will be part of the template parameters. This is not
    possible with a normal static_assert of the sizeof a type.

- **Kernel: Switch static_asserts of a type size to AK::AssertSize**

    This will provide better debug ability when the size comparison fails.

- **Userland: Switch static_assert of type sizes to AK::AssertSize**

### CODE REVIEW

File: Documentation/Patterns.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+## Introduction
+
+Overtime numerous reoccurring patterns have emerged from or where adopted by
+the serenity code base. This document aims to track and describe them so they
+can be propagated further and keep the code base consistent. 
+
+----
+## Intrusive Lists
+
+Intrusive lists are common in there Kernel and in some specific cases
```
Comment (Atari2) [inline Documentation/Patterns.md:13]: Minor typo here, `in there Kernel` -> `in the Kernel`

File: Documentation/Patterns.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    Vector<PhysicalMemoryRange> m_physical_memory_ranges;
+    Vector<ContiguousReservedMemoryRange> m_reserved_memory_ranges;
+};
+```
+
+----
+## Static Assertions of the size of a type
+
+It's a universal pattern to use `static_assert` to validate the size of a
+type matches the authors expectations. Unfortunately when these assertions
```
Comment (AtkinsSJ) [inline Documentation/Patterns.md:59]: "authors" -> "author's" (or authors')

File: Documentation/Patterns.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+fails they don't give you the values that actually caused the failure. This
```
Comment (AtkinsSJ) [inline Documentation/Patterns.md:60]: "fails" -> "fail"

File: Documentation/Patterns.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+----
+## Introduction
+
+Overtime numerous reoccurring patterns have emerged from or where adopted by
+the serenity code base. This document aims to track and describe them so they
+can be propagated further and keep the code base consistent. 
+
+----
+## Intrusive Lists
```
Comment (AtkinsSJ) [inline Documentation/Patterns.md:11]: Might be worth saying briefly what an intrusive list is and why it's useful. It's pretty easy to look up the phrase online but I hadn't heard of it before joining Serenity.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
That documentation is going to be really useful! I did find a few typos when proofreading it.

File: Documentation/Patterns.md (hunk 1 above)
Comment (bgianfo) [reply Documentation/Patterns.md]: Fixed, thanks!

File: Documentation/Patterns.md (hunk 4 above)
Comment (bgianfo) [reply Documentation/Patterns.md:11]: Sure, I added a little chunk, let me know what you think.

File: AK/StdLibExtraDetails.h [hunk 5]
```diff
@@ -475,6 +475,19 @@ using AddRvalueReference = typename __AddReference<T>::RvalueType;
...
+    static_assert(ActualSize == ExpectedSize,
+        "actual size does not match expected size");
+};
+
+// Note: This type is useful, as the sizes will be visible in the
+//       compiler error messages, as they will be part of the
+//       template parameters. This is not possible with a
+//       static_assert on the sizeof a type.
+template<typename T, unsigned ExpectedSize>
+using AssertSize = __AssertSize<T, ExpectedSize, sizeof(T)>;
```
Comment (alimpfard) [inline AK/StdLibExtraDetails.h:489]: I don't really like how you'd have to `AssertSize<foo, size>::value` to assert this, I can think of two ways, either `inline constexpr bool` for `AssertSize`, or a `consteval` function that will fail (via the same `compiletime_fail` trick used for format string checking) if the sizes aren't correct?

For comparison, here's what `static_assert(AssertSize<int, 5>::value);` produces:
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h: In instantiation of ‘struct __AssertSize<int, 5, 4>’:
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:28:33:   required from here
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: error: static assertion failed: actual size does not match expected size
   17 |     static_assert(ActualSize == ExpectedSize,
      |                   ~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: note: ‘(4 == 5)’ evaluates to false
```

what `static_assert(AssertSize<int, 5>);` (with `constexpr bool`) produces:
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h: In instantiation of ‘struct __AssertSize<int, 5, 4>’:
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:26:78:   required from ‘constexpr const bool AssertSize<int, 5>’
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:28:15:   required from here
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: error: static assertion failed: actual size does not match expected size
   17 |     static_assert(ActualSize == ExpectedSize,
      |                   ~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: note: ‘(4 == 5)’ evaluates to false
```

and here's what the most informative way I could find to make `static_assert(assert_size<int>(5));` produces:
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:46:38:   in ‘constexpr’ expansion of ‘assert_size<int>(__Actual_Size{5}, __Expected_Size{4})’
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:40:25: error: call to non-‘constexpr’ function ‘void compiletime_fail(auto:10) [with auto:10 = const char*]’
   40 |         compiletime_fail("actual size does not match expected size");
      |         ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:15:6: note: ‘void compiletime_fail(auto:10) [with auto:10 = const char*]’ declared here
   15 | void compiletime_fail(auto);
      |      ^~~~~~~~~~~~~~~~
```

File: Documentation/Patterns.md (hunk 4 above)
Comment (AtkinsSJ) [reply Documentation/Patterns.md:11]: :+1:

Though I think it should be "the elements membership" -> "the element's membership".

File: AK/StdLibExtraDetails.h (hunk 5 above)
Comment (bgianfo) [reply AK/StdLibExtraDetails.h:489]: In the interest of transparency I actually did the current and the other two examples you gave before I sent the PR. I decided that I liked the current version the most, as it was the most clear to me and gave the most debug info. I like the conteval / constexpr function, but that doesn't give you the nice compiler error message that the template type versions do.
```
/home/Test/Workspace/serenity/./AK/ByteBuffer.h:17:30: note: ‘(4 == 5)’ evaluates to false
```

So which would you prefer me to do?

File: Documentation/Patterns.md (hunk 4 above)
Comment (bgianfo) [reply Documentation/Patterns.md:11]: done

File: AK/StdLibExtraDetails.h (hunk 5 above)
Comment (alimpfard) [reply AK/StdLibExtraDetails.h:489]: my only real problem is how the assertion is spelled out, I like the `(4 == 5) evaluates to false` thing too.

Actually, one more option I can think of that probably makes me happy is:
```c++
template<typename T, unsigned ExpectedSize, unsigned ActualSize>
struct __AssertSize : TrueType {
    static_assert(ActualSize == ExpectedSize,
        "actual size does not match expected size");

    consteval operator bool() const { return value; }
};
```

so we can say `static_assert(AssertSize<int, 5>());`.

File: Documentation/Patterns.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
+# SerenityOS patterns
+
+----
```
Comment (linusg) [inline Documentation/Patterns.md:3]: not sure why we need an `<hr>` before each section?

File: Documentation/Patterns.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+element's membership in the data structure. In the case of a list, this
+means that every element in an intrusive linked list has a node embedded
+inside of it. The main advantage of intrusive
+data structures is you don't need to worry about handling out of memory (OOM)
+on insertion into the data structure. This means error handling code is
+much simpler than say, using a `Vector` in environments that need to be durable
+to OOM.
+
+The common pattern for declaring an intrusive list is to add the storage
+for the intrusive list node as a private member. A public type aliases is
```
Comment (linusg) [inline Documentation/Patterns.md:25]:
```suggestion
for the intrusive list node as a private member. A public type alias is
```

File: Documentation/Patterns.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+then used to expose the list type to anyone who might need to create it.
+Here is an example from the `Region` class in the Kernel.
```
Comment (linusg) [inline Documentation/Patterns.md:27]:
```suggestion
Here is an example from the `Region` class in the Kernel:
```

File: Documentation/Patterns.md (hunk 6 above)
Comment (bgianfo) [reply Documentation/Patterns.md:3]: Following the pattern that is is in `Documentation/SmartPointers.md`? I have no preference.

Comment (bgianfo) [reply Documentation/Patterns.md:3]: Removed since I don't really care either way.

File: AK/StdLibExtraDetails.h (hunk 5 above)
Comment (bgianfo) [reply AK/StdLibExtraDetails.h:491]: ok, done

**Pull request merged**